// uxcore-tabs/build/Tabs.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _rcTabs = require('rc-tabs');

var _rcTabs2 = _interopRequireDefault(_rcTabs);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var prefixCls = 'kuma-tab';
var STYLESUFFIX = ['underline', 'line', 'topline', 'brick'];
var SIZESUFFIX = {
	normal: '',
	mini: 'sm',
	large: 'lg'
};

var Tabs = function (_RcTabs) {
	_inherits(Tabs, _RcTabs);

	function Tabs() {
		_classCallCheck(this, Tabs);

		return _possibleConstructorReturn(this, _RcTabs.apply(this, arguments));
	}

	Tabs.prototype.render = function render() {
		var props = this.props;
		var cls = [];
		if (SIZESUFFIX[props.size]) {
			cls.push([prefixCls, SIZESUFFIX[props.size]].join('-'));
		}
		if (STYLESUFFIX.indexOf(props.tabStyle) !== -1) {
			cls.push([prefixCls, props.tabStyle].join('-'));
		}
		cls = cls.join(' ');
		return _react2["default"].createElement(_rcTabs2["default"], _extends({}, props, { className: cls }));
	};

	return Tabs;
}(_rcTabs2["default"]);

Tabs.displayName = 'uxcore-tabs';
Tabs.defaultProps = (0, _objectAssign2["default"])(_rcTabs2["default"].defaultProps, {
	prefixCls: prefixCls,
	tabStyle: 'underline',
	size: 'normal'
});
Tabs.TabPane = _rcTabs2["default"].TabPane;

exports["default"] = Tabs;
module.exports = exports['default'];
