// uxcore-dialog/build/confirm.js
'use strict';

exports.__esModule = true;

exports["default"] = function (props) {
    var div = document.createElement('div');
    document.body.appendChild(div);

    var d = void 0;
    props = props || {};
    props.iconClassName = props.iconClassName || 'kuma-icon-caution';
    var width = props.width || 416;

    props.locale = props.locale || 'zh-cn';
    var locale = _i18n2["default"][props.locale];

    // 默认为 true，保持向下兼容
    if (!('okCancel' in props)) {
        props.okCancel = true;
    }

    function close() {
        d.setState({
            visible: false
        });
        _reactDom2["default"].unmountComponentAtNode(div);
        document.body.removeChild(div);
    }

    function onCancel() {
        var cancelFn = props.onCancel;
        if (cancelFn) {
            var ret = void 0;
            if (cancelFn.length) {
                ret = cancelFn(close);
            } else {
                ret = cancelFn();
                if (!ret) {
                    close();
                }
            }
            if (ret && ret.then) {
                ret.then(close);
            }
        } else {
            close();
        }
    }

    function onOk() {
        var okFn = props.onOk;
        if (okFn) {
            var ret = void 0;
            if (okFn.length) {
                ret = okFn(close);
            } else {
                ret = okFn();
                if (!ret) {
                    close();
                }
            }
            if (ret && ret.then) {
                ret.then(close);
            }
        } else {
            close();
        }
    }

    var body = _react2["default"].createElement(
        'div',
        { className: 'kuma-confirm-body' },
        _react2["default"].createElement('i', { className: 'kuma-icon ' + props.iconClassName }),
        _react2["default"].createElement(
            'span',
            { className: 'kuma-confirm-title' },
            props.title
        ),
        _react2["default"].createElement(
            'div',
            { className: 'kuma-confirm-content' },
            props.content
        )
    );
    var footer = void 0;

    if (props.okCancel) {
        footer = _react2["default"].createElement(
            'div',
            { className: 'kuma-confirm-action' },
            _react2["default"].createElement(
                _uxcoreButton2["default"],
                { size: props.buttonSize || "medium", onClick: onOk },
                locale['ok']
            ),
            _react2["default"].createElement(
                _uxcoreButton2["default"],
                { type: 'secondary', size: props.buttonSize || "medium", onClick: onCancel },
                locale['cancel']
            )
        );
    } else {
        footer = _react2["default"].createElement(
            'div',
            { className: 'kuma-confirm-action' },
            _react2["default"].createElement(
                _uxcoreButton2["default"],
                { size: props.buttonSize || "medium", onClick: onOk },
                locale['isee']
            )
        );
    }

    _reactDom2["default"].render(_react2["default"].createElement(
        _Dialog2["default"],
        {
            prefixCls: 'kuma-dlg',
            className: 'kuma-dlg-confirm',
            visible: true,
            closable: false,
            title: '',
            footer: '',
            maskTransitionName: 'fade', width: width },
        _react2["default"].createElement(
            'div',
            null,
            body,
            ' ',
            footer
        )
    ), div, function () {
        d = this;
    });
};

var _uxcoreButton = require('uxcore-button');

var _uxcoreButton2 = _interopRequireDefault(_uxcoreButton);

var _Dialog = require('./Dialog');

var _Dialog2 = _interopRequireDefault(_Dialog);

var _i18n = require('./i18n');

var _i18n2 = _interopRequireDefault(_i18n);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

module.exports = exports['default'];
