// uxcore-dialog/build/Dialog.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _rcDialog = require('rc-dialog');

var _rcDialog2 = _interopRequireDefault(_rcDialog);

var _confirm = require('./confirm');

var _confirm2 = _interopRequireDefault(_confirm);

var _uxcoreButton = require('uxcore-button');

var _uxcoreButton2 = _interopRequireDefault(_uxcoreButton);

var _i18n = require('./i18n');

var _i18n2 = _interopRequireDefault(_i18n);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author: vincent
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @date: 15/5/19
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


function noop() {}

var Dialog = function (_React$Component) {
    _inherits(Dialog, _React$Component);

    function Dialog(props) {
        _classCallCheck(this, Dialog);

        var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

        _this.state = {
            confirmLoading: false,
            visible: props.visible
        };
        return _this;
    }

    Dialog.prototype.handleCancel = function handleCancel() {
        this.props.onCancel();
    };

    Dialog.prototype.handleOk = function handleOk() {
        this.props.onOk();
    };

    Dialog.prototype.render = function render() {
        var props = this.props;
        var locale = _i18n2["default"][props.locale];
        var defaultFooter = [_react2["default"].createElement(
            _uxcoreButton2["default"],
            { key: 'confirm',
                type: 'primary',
                size: 'medium',
                loading: props.confirmLoading,
                onClick: this.handleOk.bind(this) },
            locale['ok']
        ), _react2["default"].createElement(
            _uxcoreButton2["default"],
            { key: 'cancel',
                type: 'secondary',
                size: 'medium',
                onClick: this.handleCancel.bind(this) },
            locale['cancel']
        )];
        var footer = props.footer || defaultFooter;
        return _react2["default"].createElement(_rcDialog2["default"], _extends({ onClose: this.handleCancel.bind(this), footer: footer }, props, {
            visible: props.visible }));
    };

    return Dialog;
}(_react2["default"].Component);

exports["default"] = Dialog;


Dialog.defaultProps = {
    prefixCls: 'kuma-dlg',
    onOk: noop,
    locale: 'zh-cn',
    onCancel: noop,
    width: 520,
    transitionName: '',
    maskAnimation: false,
    confirmLoading: false,
    visible: false
};

Dialog.info = function (props) {
    props.iconClassName = 'kuma-icon-information';
    props.okCancel = false;
    return (0, _confirm2["default"])(props);
};

Dialog.success = function (props) {
    props.iconClassName = 'kuma-icon-success';
    props.okCancel = false;
    return (0, _confirm2["default"])(props);
};

Dialog.error = function (props) {
    props.iconClassName = 'kuma-icon-error';
    props.okCancel = false;
    return (0, _confirm2["default"])(props);
};

Dialog.confirm = function (props) {
    props.okCancel = true;
    return (0, _confirm2["default"])(props);
};
module.exports = exports['default'];
