// uxcore-form/build/FormField/TextAreaFormField.js
'use strict';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Created by xy on 15/4/13.
 */
var FormField = require('./FormField');
var Constants = require("uxcore-const");
var assign = require('object-assign');

var TextAreaFormField = function (_FormField) {
    _inherits(TextAreaFormField, _FormField);

    function TextAreaFormField(props) {
        _classCallCheck(this, TextAreaFormField);

        return _possibleConstructorReturn(this, _FormField.call(this, props));
    }

    TextAreaFormField.prototype.handleChange = function handleChange(e) {
        var me = this;
        var autoTrim = me.props.autoTrim;

        var value = e.currentTarget.value;
        if (autoTrim) {
            value = me.trim(value);
        }
        me.handleDataChange(value);
    };

    TextAreaFormField.prototype.trim = function trim(str) {
        return str.replace(/(^\s+|\s+$)/g, "");
    };

    TextAreaFormField.prototype.addSpecificClass = function addSpecificClass() {
        var me = this;
        if (me.props.jsxprefixCls == "kuma-uxform-field") {
            return me.props.jsxprefixCls + " kuma-textarea-uxform-field";
        } else {
            return me.props.jsxprefixCls;
        }
    };

    TextAreaFormField.prototype.handleFocus = function handleFocus(e) {
        this.props.onFocus(e);
    };

    TextAreaFormField.prototype.handleBlur = function handleBlur(e) {
        var me = this;
        var pass = true;
        if (me.props.validateOnBlur) {
            pass = me.doValidate();
        }
        me.props.onBlur(e, pass);
    };

    TextAreaFormField.prototype.handleKeyDown = function handleKeyDown(e) {
        var me = this;
        me.props.onKeyDown(e);
    };

    TextAreaFormField.prototype.renderField = function renderField() {
        var me = this;
        var mode = me.props.jsxmode || me.props.mode;
        if (mode == Constants.MODE.EDIT) {
            return React.createElement('textarea', {
                disabled: me.props.jsxdisabled,
                placeholder: me.props.jsxplaceholder,
                className: 'kuma-textarea',
                ref: 'root',
                value: me.state.value || "",
                onChange: me.handleChange.bind(me),
                onFocus: me.handleFocus.bind(me),
                onBlur: me.handleBlur.bind(me),
                onKeyDown: me.handleKeyDown.bind(me) });
        } else if (mode == Constants.MODE.VIEW) {
            return React.createElement(
                'span',
                null,
                me.state.value
            );
        }
    };

    return TextAreaFormField;
}(FormField);

TextAreaFormField.displayName = "TextAreaFormField";
TextAreaFormField.propTypes = assign({}, FormField.propTypes, {
    onBlur: React.PropTypes.func,
    onFocus: React.PropTypes.func,
    onKeyDown: React.PropTypes.func,
    validateOnBlur: React.PropTypes.bool,
    autoTrim: React.PropTypes.bool
});
TextAreaFormField.defaultProps = assign({}, FormField.defaultProps, {
    onBlur: function onBlur() {},
    onFocus: function onFocus() {},
    onKeyDown: function onKeyDown() {},
    validateOnBlur: false
});
module.exports = TextAreaFormField;
