// uxcore-table/build/Cell/RadioField.js
'use strict';

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var CellField = require('./CellField');
var classnames = require('classnames');
var assign = require('object-assign');
var RadioGroup = require('uxcore-radiogroup');
var React = require('react');
var ReactDOM = require('react-dom');

var RadioField = function (_CellField) {
    _inherits(RadioField, _CellField);

    function RadioField(props) {
        _classCallCheck(this, RadioField);

        return _possibleConstructorReturn(this, _CellField.call(this, props));
    }

    RadioField.prototype.processChildren = function processChildren() {
        var me = this;
        var obj = {};
        me.props.column.children.forEach(function (item) {
            obj[item.props.value] = item.props.text;
        });
        return obj;
    };

    RadioField.prototype.renderContent = function renderContent() {
        var me = this;
        var dataKey = me.props.column.dataKey;
        var textMap = me.processChildren();
        var fieldProps = {
            onChange: function onChange(value) {
                me.handleDataChange({
                    jsxid: me.props.rowData['jsxid'],
                    column: me.props.column,
                    text: textMap[value],
                    value: value
                });
            },
            value: me.props.value
        };
        if (me.props.column.config) {
            var _me$props$column$conf = me.props.column.config;
            var value = _me$props$column$conf.value;
            var onChange = _me$props$column$conf.onChange;

            var customProps = _objectWithoutProperties(_me$props$column$conf, ['value', 'onChange']);

            assign(fieldProps, customProps);
        }
        return React.createElement(
            RadioGroup,
            fieldProps,
            me.props.column.children
        );
    };

    return RadioField;
}(CellField);

;

RadioField.propTypes = assign({}, CellField.propTypes);

RadioField.defaultProps = assign({}, CellField.defaultProps);

module.exports = RadioField;
