// uxcore-table/build/Cell/SelectField.js
'use strict';

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var CellField = require('./CellField');
var classnames = require('classnames');
var assign = require('object-assign');
var Select = require('uxcore-select2');
var Option = Select.Option;

var React = require('react');
var ReactDOM = require('react-dom');

var SelectField = function (_CellField) {
    _inherits(SelectField, _CellField);

    function SelectField(props) {
        _classCallCheck(this, SelectField);

        return _possibleConstructorReturn(this, _CellField.call(this, props));
    }

    SelectField.prototype.renderContent = function renderContent() {
        var me = this;
        var fieldProps = {
            onSelect: function onSelect(value, Option) {
                me.handleDataChange({
                    jsxid: me.props.rowData['jsxid'],
                    column: me.props.column,
                    text: Option.props.children,
                    value: value
                });
            },
            value: me.props.value
        };
        if (me.props.column.config) {
            var _me$props$column$conf = me.props.column.config;
            var value = _me$props$column$conf.value;
            var onSelect = _me$props$column$conf.onSelect;

            var customProps = _objectWithoutProperties(_me$props$column$conf, ['value', 'onSelect']);

            assign(fieldProps, customProps);
        }
        return React.createElement(
            Select,
            fieldProps,
            me.props.column.children
        );
    };

    return SelectField;
}(CellField);

;

SelectField.propTypes = assign({}, CellField.propTypes);

SelectField.defaultProps = assign({}, CellField.defaultProps);

module.exports = SelectField;
