// uxcore-calendar/build/YearCalendar.js
'use strict';

exports.__esModule = true;

var _YearPanel = require('rc-calendar/lib/year/YearPanel');

var _YearPanel2 = _interopRequireDefault(_YearPanel);

var _CalendarMixin = require('rc-calendar/lib/mixin/CalendarMixin');

var _CalendarMixin2 = _interopRequireDefault(_CalendarMixin);

var _rcUtil = require('rc-util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var React = require('react');
var ReactDOM = require('react-dom');

var YearCalendar = React.createClass({
    displayName: 'YearCalendar',

    mixins: [_CalendarMixin2["default"]],

    onKeyDown: function onKeyDown(e) {
        var keyCode = e.keyCode;
        var ctrlKey = e.ctrlKey || e.metaKey;
        var stateValue = this.state.value;
        var value = stateValue;
        switch (keyCode) {
            case _rcUtil.KeyCode.DOWN:
                value = stateValue.clone();
                value.addMonth(3);
                break;
            case _rcUtil.KeyCode.UP:
                value = stateValue.clone();
                value.addMonth(-3);
                break;
            case _rcUtil.KeyCode.LEFT:
                value = stateValue.clone();
                if (ctrlKey) {
                    value.addYear(-1);
                } else {
                    value.addMonth(-1);
                }
                break;
            case _rcUtil.KeyCode.RIGHT:
                value = stateValue.clone();
                if (ctrlKey) {
                    value.addYear(1);
                } else {
                    value.addMonth(1);
                }
                break;
            case _rcUtil.KeyCode.ENTER:
                this.onSelect(stateValue);
                e.preventDefault();
                return 1;
            default:
                return undefined;
        }
        if (value !== stateValue) {
            this.setValue(value);
            e.preventDefault();
            return 1;
        }
    },
    render: function render() {
        var props = this.props;
        var children = React.createElement(_YearPanel2["default"], { locale: props.locale,
            disabledDate: props.disabledDate,
            value: this.state.value,
            rootPrefixCls: props.prefixCls,
            onChange: this.setValue,
            onSelect: this.onSelect });
        return this.renderRoot({
            children: children
        });
    }
});

exports["default"] = YearCalendar;
module.exports = exports['default'];
