// uxcore-uploader/build/FileList.js
'use strict';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var FileItem = require('./FileItem');
var Picker = require('./Picker');

var _require = require('uploadcore');

var Events = _require.Events;

var React = require('react');
var ReactDOM = require('react-dom');

var FileList = function (_React$Component) {
    _inherits(FileList, _React$Component);

    function FileList(props) {
        _classCallCheck(this, FileList);

        var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

        _this.core = _this.props.core;

        _this.state = {
            items: _this.core.getStat().getFiles()
        };
        return _this;
    }

    FileList.prototype.componentDidMount = function componentDidMount() {
        var _this2 = this;

        var statchange = function statchange(stat) {
            _this2.setState({
                items: stat.getFiles()
            });
        };
        this.core.on(Events.QUEUE_STAT_CHANGE, statchange);
        this.stopListen = function () {
            _this2.core.off(Events.QUEUE_STAT_CHANGE, statchange);
        };
    };

    FileList.prototype.componentWillUnmount = function componentWillUnmount() {
        this.stopListen && this.stopListen();
    };

    FileList.prototype.render = function render() {
        var _this3 = this;

        return React.createElement(
            'div',
            { className: "kuma-upload-filelist " + (this.props.mode === 'nw' ? 'nwmode' : this.props.mode === 'mini' ? 'minimode' : 'iconmode') },
            React.createElement(
                'div',
                { className: 'inner' },
                this.state.items.map(function (file) {
                    return React.createElement(FileItem, { locale: _this3.props.locale, key: file.id, file: file, mode: _this3.props.mode });
                }),
                !this.core.isFull() && this.props.mode === 'icon' ? React.createElement(
                    Picker,
                    { core: this.core },
                    React.createElement('i', { className: 'kuma-icon kuma-icon-add' })
                ) : null
            )
        );
    };

    return FileList;
}(React.Component);

FileList.defaultProps = {
    mode: 'mini'
};

module.exports = FileList;
